pkg/admission/namespacelifecycle/admission.go

```diff
@@ -0,0 +1,161 @@
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clusters"
+)
+
+const (
+	// PluginName indicates the name of admission plug-in
+	PluginName = "WorkspaceNamespaceLifecycle"
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return newLifcycle()
+	})
+}
+
+// workspaceNamespaceLifecycle is a wrapper of kubernetes namespace lifecycle admission control
+// It uses legacy namespace lifecycle admission control by default, and ignores
+// immortal namespaces when the workspaces is deleting. This can ensure we can remove all immortal
+// namespaces when workspaces is deleting.
+type workspaceNamespaceLifecycle struct {
+	// legacyNamespaceLifecycle is the kube legacy namespace lifecycle
+	legacyNamespaceLifecycle *lifecycle.Lifecycle
+
+	// namespaceLifecycle is used only when workspace is deleting
+	namespaceLifecycle *lifecycle.Lifecycle
+
+	*admission.Handler
+	workspaceLister tenancyv1alpha1lister.ClusterWorkspaceLister
+}
+
+func newLifcycle() (*workspaceNamespaceLifecycle, error) {
+	legacyLifecycle, err := lifecycle.NewLifecycle(sets.NewString(metav1.NamespaceDefault, metav1.NamespaceSystem, metav1.NamespacePublic))
+	if err != nil {
+		return nil, err
+	}
+
+	lifecycle, err := lifecycle.NewLifecycle(sets.NewString())
+	if err != nil {
+		return nil, err
+	}
+
+	return &workspaceNamespaceLifecycle{
+		Handler:                  admission.NewHandler(admission.Create, admission.Update, admission.Delete),
+		legacyNamespaceLifecycle: legacyLifecycle,
+		namespaceLifecycle:       lifecycle,
+	}, nil
+}
+
+var _ = kcpinitializers.WantsKcpInformers(&workspaceNamespaceLifecycle{})
+var _ = initializer.WantsExternalKubeInformerFactory(&workspaceNamespaceLifecycle{})
+var _ = initializer.WantsExternalKubeClientSet(&workspaceNamespaceLifecycle{})
+
+// Admit makes an admission decision based on the request attributes
+func (l *workspaceNamespaceLifecycle) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+
+	// call legacy namespace lifecycle at first
+	admissionErr := l.legacyNamespaceLifecycle.Admit(ctx, a, o)
+
+	// Only handle the case when namespace deletion is forbidden
+	if !apierrors.IsForbidden(admissionErr) {
+		return admissionErr
+	}
+
+	if a.GetOperation() != admission.Delete || a.GetKind().GroupKind() != corev1.SchemeGroupVersion.WithKind("Namespace").GroupKind() {
+		return admissionErr
+	}
+
+	clusterName, err := genericapirequest.ClusterNameFrom(ctx)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	org, hasParent := clusterName.Parent()
+	if !hasParent {
+		return admissionErr
+	}
+
+	workspaceKey := clusters.ToClusterAwareKey(org, clusterName.Base())
+
+	workspace, err := l.workspaceLister.Get(workspaceKey)
+	// The shard hosting the workspace could done,
+	// just return error from legacy namespace lifecycle addmission in this case
+	if err != nil {
```

Comment:
do we want to handle 404 differently? A 404 is a definitive response, i.e. the workspace is really gone. Maybe we want to keep deleting the namespace then?